lstchain/image/modifier.py

```diff
@@ -577,3 +537,160 @@ def calculate_required_additional_nsb(simtel_filename, data_dl1_filename, config
     extra_nsb = ((data_ped_variance - mc_ped_variance)
                  / mc_ped_variance)
     return extra_nsb, data_ped_variance, mc_ped_variance
+
+
+class WaveformNsbTunner:
+    """
+    Handles the injection of additional NSB pulses in waveforms.
+    """
+    def __init__(self, added_nsb_fraction, original_nsb, pulse_template, charge_spe_cumulative_pdf,
+                 pre_computed_multiplicity=10):
+        """
+        Parameters
+        ----------
+        added_nsb_fraction: `float`
+            Fraction of `original_nsb` to be added during NSB injection. Can be more than 1.
+        original_nsb: `dict`[`int`,`astropy.units.Quantity`]
+            NSB frequency used in the original waveforms per telescope
+        pulse_template: `dict`[int,`lstchain.data.NormalizedPulseTemplate`]
+            Single photo-electron pulse template per telescope
+        charge_spe_cumulative_pdf: `scipy.interpolate.interp1d`
+            Cumulative amplitude probability density function for single photo-electrons
+        pre_computed_multiplicity: `int`
+            Multiplicative factor on the number of pixels used to determine the number of pre-generated, nsb-only,
+            waveforms. Later used during event modification. Set to 0 to always compute the correction on the fly.
+
+        """
+        self.added_nsb_fraction = added_nsb_fraction
+        self.original_nsb = original_nsb
+        self.pulse_template = pulse_template
+        self.charge_spe_cumulative_pdf = charge_spe_cumulative_pdf
+        self.multiplicity = pre_computed_multiplicity
+        self.nsb_waveforms = {}
+        self.nb_simulated = {}
+        self.rng = np.random.default_rng()
+        if self.multiplicity == 0:
+            self.tune_nsb_on_waveform = self._tune_nsb_on_waveform_direct
+        else:
+            self.tune_nsb_on_waveform = self._tune_nsb_on_waveform_precomputed
+
+    def initialise_waveforms(self, waveform, dt, tel_id):
+        """
+        Creates an array of nsb only waveforms later used to inject additional nsb in events.
+        Called once for each telescope id at first encounter in `self._tune_nsb_on_waveform_precomputed`.
+
+        Parameters
+        ----------
+        waveform: ndarray
+            Waveform used to know the number of pixels and time samples for a given telescope id
+        dt: `astropy.units.Quantity`
+            Time between samples
+        tel_id: `int`
+            Telescope identifier for which nsb waveforms are generated
+
+        """
+        log.info(f"Pre-generating nsb waveforms for nsb tuning and telescope id {tel_id}.")
+        n = 25
```

Comment:
this n = 25 is defined several times, does it make sense to define it once at the beginning of the file?

Reply:
It does, I will change it.